tektoncd-cli: cleanup

06kellyjac opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

- use buildGoModule buildPhase
- let buildGoModule set platform
- add -s -w ldflags for smaller build (-5MB)
  - 43MB -> 38MB
- add checkPhase running unit tests
- install man pages in 1 line
- add changelog
- add @vdemeester as a maintainer

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

06kellyjac commented:
Result of `nixpkgs-review pr 112755` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tektoncd-cli</li>
  </ul>
</details>

vdemeester commented on pkgs/applications/networking/cluster/tektoncd-cli/default.nix:
```diff
@@ -13,35 +13,39 @@ buildGoModule rec {
-  buildPhase = ''
-    make bin/tkn
+  checkPhase = ''
+    make test-unit
```
🎉

vdemeester replied:
Although, I think relying on `go test` (that `buildGoModule` does automatically right ?) should work fine.

06kellyjac replied:
Leaving `checkPhase` to `buildGoModule` said no tests found as it only looks in `cmd/tkn`

```
running tests
?       github.com/tektoncd/cli/cmd/tkn [no test files]
```

I'll try some different `subPackages` options

06kellyjac replied:
If you don't set `subPackages`:

```
Building subPackage ./third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap
main module (github.com/tektoncd/cli) does not contain package github.com/tektoncd/cli/third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap
error: --- Error
```

06kellyjac replied:
`rm -rf third_party` lets it run pretty well. Hitting a couple failing tests with the version

Should I sed `Client version: dev` -> `Client version: 0.16.0` ?

Alternatively I could update the golden files then testing it matches the golden files, but that seems useless

```
=== RUN   TestGetVersions/empty_deployment_items
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: unknown, pipeline controller may be installed in another namespace please use tkn version -n {namespace}



        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-empty_deployment_items.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_only_with_pipeline_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0



        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_only_with_pipeline_installed.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_with_pipeline_and_triggers_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0
         Triggers version: v0.5.0


        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_with_pipeline_and_triggers_installed.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_with_pipeline_and_dashboard_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0
         Dashboard version: v0.7.0


        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_with_pipeline_and_dashboard_installed.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_with_pipeline,_triggers_and_dashboard_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0
         Triggers version: v0.5.0


        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_with_pipeline,_triggers_and_dashboard_installed.golden to the new expected value.'

--- FAIL: TestGetVersions (0.00s)
    --- FAIL: TestGetVersions/empty_deployment_items (0.00s)
    --- FAIL: TestGetVersions/deployment_only_with_pipeline_installed (0.00s)
    --- FAIL: TestGetVersions/deployment_with_pipeline_and_triggers_installed (0.00s)
    --- FAIL: TestGetVersions/deployment_with_pipeline_and_dashboard_installed (0.00s)
    --- FAIL: TestGetVersions/deployment_with_pipeline,_triggers_and_dashboard_installed (0.00s)
```

06kellyjac replied:
That worked, all tests passed. I'll commit and push. LMK if there's a cleaner way to do this :)

06kellyjac replied:
Would it be worth `rm -rf` all the .go files in VENDOR-LICENSE in https://github.com/tektoncd/cli?
Should I post a bug report in https://github.com/google/go-license?
Or is including the .go files a requirement of the license? Looks like it's all just MPL-2.0 License though.

```
λ find ./third_party/VENDOR-LICENSE -type f -name "*.go"
./third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap/errwrap.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap/errwrap_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/2q_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/2q.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/simplelru/lru.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/simplelru/lru_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/simplelru/lru_interface.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/lru.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/arc.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/doc.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/lru_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/arc_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/multierror_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/sort_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/prefix.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/flatten_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/append.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/group.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/multierror.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/format_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/sort.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/prefix_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/flatten.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/append_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/format.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/group_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version_collection_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/constraint.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version_collection.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/constraint_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version_test.go
```

vdemeester replied:
oh I didn't think of that 😅 (both for the subpackage and the tests that check version)

vdemeester replied:
I think it looks good 😉

06kellyjac replied:
:+1: If these files are removed from https://github.com/tektoncd/cli at some point we can drop the line

06kellyjac replied:
@vdemeester Not sure where else to ask but is it possible to change the release note of a merged PR before release?

vdemeester reviewed (approved):
LGTM 🐯

r-rmcgibbo commented:
Result of `nixpkgs-review pr 112755` at 3f5d12f2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/849e6b241ca823c8be2b22114e31f715">tektoncd-cli</a></li></li>
  </ul>
</details>

vdemeester commented:
LGTM 🐯

06kellyjac commented:
Result of `nixpkgs-review pr 112755` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tektoncd-cli</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/applications/networking/cluster/tektoncd-cli/default.nix:
```diff
@@ -13,35 +13,43 @@ buildGoModule rec {
-    make bin/tkn
+  preBuild = ''
+    # Breaks build/check as go files are still included
+    rm -rf ./third_party/VENDOR-LICENSE
```
```suggestion
    rm -r ./third_party/VENDOR-LICENSE
```

SuperSandro2000 commented on pkgs/applications/networking/cluster/tektoncd-cli/default.nix:
```diff
@@ -13,35 +13,43 @@ buildGoModule rec {
+  '';
+
+  postInstall = ''
+    install -D -m 644 docs/man/man1/* -t $out/man/man1
```
Can we use installManPage here?

06kellyjac replied:
I think that's worked
```
 result ⇒ /nix/store/sz1dx4w37g1f07prnvfmsf3d1lzxiqqk-tektoncd-cli-0.16.0
├──  bin
│  └──  tkn
└──  share
   ├──  bash-completion
   │  └──  completions
   │     └──  tkn.bash
   ├──  fish
   │  └──  vendor_completions.d
   │     └──  tkn.fish
   ├──  man
   │  └──  man1
   │     ├──  tkn-clustertask-delete.1.gz
   │     ├──  tkn-clustertask-describe.1.gz
```

06kellyjac replied:
Just tweaking before pushing

SuperSandro2000 commented on pkgs/applications/networking/cluster/tektoncd-cli/default.nix:
```diff
@@ -13,35 +13,47 @@ buildGoModule rec {
   installPhase = ''
-    install bin/tkn -Dt $out/bin
+    # Only install tkn
+    install -D $GOPATH/bin/tkn $out/bin/tkn
```
Not sure about the $GOPATH here. Why did you change that?

kalbasit replied:
I think we should use `subPackages` and use postInstall to install completions and man pages rather than overriding the installPhase IMO.

06kellyjac replied:
Well is there an easy way to filter out the relevant subPackages? cmd/tkn obviously but that has none of the tests

06kellyjac replied:
Ok I'm excluding `third_party` & `cmd/docs` so now I don't need to rm third_party or get rid of cmd/docs result

06kellyjac replied:
The interface for `excludePackages` is really uncomfortable. It'd be nice if it could take an array like `subPackages`

kalbasit commented on pkgs/applications/networking/cluster/tektoncd-cli/default.nix:
```diff
@@ -13,35 +13,42 @@ buildGoModule rec {
       --bash <($out/bin/tkn completion bash) \
       --fish <($out/bin/tkn completion fish) \
-      --zsh <($out/bin/tkn completion zsh)
+      --zsh  <($out/bin/tkn completion zsh)
```
Please revert this change which adds an extra unnecessary space.

06kellyjac committed:
tektoncd-cli: cleanup

- use buildGoModule buildPhase
- let buildGoModule set platform
- add -s -w ldflags for smaller build (-5MB)
- add checkPhase running unit tests
- install man pages in 1 line
- add changelog
- add @vdemeester as a maintainer

Mic92 commented:
Result of `nixpkgs-review pr 112755` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tektoncd-cli</li>
  </ul>
</details>
